### PR TITLE
Refactor rendering process - use Jinja templates

### LINKED ISSUE #187: Refactor rendering process to use more JINJA templates (0/1)

### PR DESCRIPTION
**Pull request by mik-laj**
Hello,

I have rewritten the workflow generation process to use more Jinja templates. Additionally, I completed the tests.

It's a part of https://github.com/GoogleCloudPlatform/cloud-composer/issues/162

Thanks,

Closes #187

### CODE REVIEW

**Comment by potiuk** [comment]
I propose you with that @mik-laj until I release and we merge the change with processing wf:functions. It changes quite a bit the way templates are being rendered and I would like to avoid having to rewrite it again until I finish. So I propose do not invest any time on it @mik-laj until my change is reviewed and merged.

**Comment by mik-laj** [comment]
I did not split the Converter class into several classes, because it would be difficult to do in one commit. I decided that the whole code is clear enough. Now the class is much clearer.

File: oozie-to-airflow/converter/parsed_node.py [hunk 1]
```diff
@@ -22,17 +22,25 @@
...
+from converter import primitives  # noqa: F401 pylint: disable=unused-import
 
 
 class ParsedNode:
     """Class for parsed Oozie workflow node"""
 
-    def __init__(self, mapper: "base_mapper.BaseMapper"):
+    def __init__(
+        self,
+        mapper: "base_mapper.BaseMapper",
```
Comment (potiuk) [inline oozie-to-airflow/converter/parsed_node.py]: We should use classes here not strings. That makes it easier for IntelliJ usage and it makes use of import statements which is good to show that there are no hidden dependencies.

In case it causes a circular dependency, then it should be fixed in the structure or (worst case) we can use local imports in this case.

File: oozie-to-airflow/converter/primitives.py [hunk 2]
```diff
@@ -42,17 +42,25 @@ class Workflow:  # pylint: disable=too-few-public-methods
     nodes: Dict[str, "parsed_node.ParsedNode"]
     dependencies: Set[str]  # TODO: Check is set likely maintain insertion order (Python 3.6 ?)
```
Comment (potiuk) [inline oozie-to-airflow/converter/primitives.py]: Those should be removed as discussed in #180 . But we can do it in a separate PR /refactor most likely.

File: pylintrc [hunk 3]
```diff
@@ -144,7 +144,8 @@ disable=print-statement,
         logging-fstring-interpolation,
         import-error,
         duplicate-code,
-        fixme
+        fixme,
+        cyclic-import
```
Comment (potiuk) [inline pylintrc]: We should not remove cyclic-import

**Comment by potiuk** [comment]
One more comment. It's not part of #162 - I created a missing issue: #187 . Please make sure next time @mik-laj to create appropriate issue first and add it to project as we are now moving to more github-issues organized process.

File: oozie-to-airflow/converter/oozie_converter.py [hunk 4]
```diff
@@ -93,123 +92,76 @@ def __init__(
...
-        file.write("PARAMS = " + json.dumps(converted_params, indent=INDENT) + "\n\n")
-
-    def write_nodes(self, file: TextIO, nodes: Dict[str, ParsedNode], indent: int = INDENT):
-        """
-        Writes the Airflow tasks to the given opened file object.
-
-        :param file: The file pointer to write to.
-        :param nodes: Dictionary of {'task_id', ParsedNode}
-        :param indent: integer of how many spaces to indent entire operator
+        Copies additional assets needed to execute a workflow, eg Pig scripts
```
Comment (sprzedwojski) [inline oozie-to-airflow/converter/oozie_converter.py]:
```suggestion
        Copies additional assets needed to execute a workflow, eg. Pig scripts.
```

File: oozie-to-airflow/converter/oozie_converter.py [hunk 5]
```diff
@@ -93,123 +92,76 @@ def __init__(
...
-        These are each written on a new line.
+    def render_workflow(self, workflow: Workflow):
         """
-        logging.info("Writing control flow dependencies to file.")
-        relations_str = render_template(template_name="relations.tpl", relations=relations)
-        file.write(textwrap.indent(relations_str, indent * " "))
-
-    @staticmethod
-    def write_dependencies(file, depends, line_prefix=""):
+        Creates text representation of the workflow
```
Comment (sprzedwojski) [inline oozie-to-airflow/converter/oozie_converter.py]:
```suggestion
        Creates text representation of the workflow.
```

File: oozie-to-airflow/templates/subworkflow.tpl [hunk 6]
```diff
@@ -14,8 +14,21 @@
   limitations under the License.
  #}
 
+{%- for depedency in dependencies %}
```
Comment (sprzedwojski) [inline oozie-to-airflow/templates/subworkflow.tpl]:
```suggestion
{%- for dependency in dependencies %}
```

File: oozie-to-airflow/templates/subworkflow.tpl [hunk 7]
```diff
@@ -14,8 +14,21 @@
...
+{{ depedency }}
```
Comment (sprzedwojski) [inline oozie-to-airflow/templates/subworkflow.tpl]:
```suggestion
{{ dependency }}
```

File: oozie-to-airflow/converter/oozie_converter.py [hunk 8]
```diff
@@ -93,123 +92,76 @@ def __init__(
...
-        os.makedirs(self.output_directory_path, exist_ok=True)
+        workflow = self.parser.workflow
+        self.convert_nodes(workflow.nodes)
+        self.create_dag_file(workflow)
+        self.copy_extra_assets(workflow.nodes)
+
+    @staticmethod
+    def convert_nodes(nodes: Dict[str, ParsedNode]):
+        """
+        For each oozie node, converts it into relations and internal relations
```
Comment (sprzedwojski) [inline oozie-to-airflow/converter/oozie_converter.py]:
```suggestion
        For each Oozie node, converts it into relations and internal relations.
```

File: oozie-to-airflow/templates/workflow.tpl [hunk 9]
```diff
@@ -14,8 +14,18 @@
   limitations under the License.
  #}
 
+{%- for depedency in dependencies %}
```
Comment (sprzedwojski) [inline oozie-to-airflow/templates/workflow.tpl]:
```suggestion
{%- for dependency in dependencies %}
```

File: oozie-to-airflow/templates/workflow.tpl [hunk 10]
```diff
@@ -14,8 +14,18 @@
...
+{{ depedency }}
```
Comment (sprzedwojski) [inline oozie-to-airflow/templates/workflow.tpl]:
```suggestion
{{ dependency }}
```

File: oozie-to-airflow/converter/oozie_converter.py (hunk 4 above)
Comment (potiuk) [reply oozie-to-airflow/converter/oozie_converter.py]: I think we must have linter and grammar check indeed :). I will add issue for that :)

File: oozie-to-airflow/converter/oozie_converter.py (hunk 5 above)
Comment (mik-laj) [reply oozie-to-airflow/converter/oozie_converter.py]: I updated the PR.

File: oozie-to-airflow/templates/subworkflow.tpl (hunk 6 above)
Comment (mik-laj) [reply oozie-to-airflow/templates/subworkflow.tpl]: I updated the PR.

File: oozie-to-airflow/templates/subworkflow.tpl (hunk 7 above)
Comment (mik-laj) [reply oozie-to-airflow/templates/subworkflow.tpl]: I updated the PR.

File: oozie-to-airflow/converter/oozie_converter.py (hunk 8 above)
Comment (mik-laj) [reply oozie-to-airflow/converter/oozie_converter.py]: I updated the PR. Thanks.

File: oozie-to-airflow/templates/workflow.tpl (hunk 9 above)
Comment (mik-laj) [reply oozie-to-airflow/templates/workflow.tpl]: I updated the PR. Thanks a lot.

File: oozie-to-airflow/templates/workflow.tpl (hunk 10 above)
Comment (mik-laj) [reply oozie-to-airflow/templates/workflow.tpl]: I updated the PR. Thanks a lot. You're awesome.
